docs/architecture/adr-001-redesign-knuu-async-approach.md

```diff
@@ -0,0 +1,49 @@
+# ADR 001: Redesigning Knuu for Asynchronous Testing Approach
+
+## Status
+
+Draft
+
+## Context
+
+The existing design of `Knuu` requires exposing the addresses and ports of the target instance running alongside BitTwister, complicating the configuration process. This setup requires Knuu to orchestrate the instances, configure them, and execute the test, leading to complexities in managing traffic shaping. Moreover, as the current model relies on port forwarding, scalability issues have surfaced, causing inconsistent errors and hindering seamless test execution.
+
+## Problem
+
+The current synchronous model necessitates exposing the target instance's address and ports to BitTwister via a sidecar. This centralized approach introduces dependencies and complexities in configuring the test environment. Additionally, the reliance on a proxy approach or port forwarding poses scalability issues, resulting in inconsistent errors during test execution.
+
+## Proposed Solution
+
+### A Distributed Approach with a Playbook Model
+
+Embrace a distributed playbook model similar to Testground, allowing asynchronous orchestration and execution of test scenarios:
+
+1. **Test Scenario Definition:** Users define test scenarios in a descriptive language (e.g., Golang).
+2. **Knuu Instance Initiation:** Knuu initiates instances and assigns specific playbooks to each instance.
+3. **BitTwister Sidecar with Playbook Understanding:** BitTwister runs alongside instances with added functionality to interpret and execute the playbook's instructions.
```

Comment:
I would propose decoupling the playbook logic from bittwister, as this playbook can have other use cases such as injecting transactions, restarting the node, etc.

Reply:
Good ideas, thanks guys. To have things clean and reduce communication overhead, we can embed BitTwister code inside that playbook. It can simply import it and control it according to the given plan. I do agree to keep its code base separate and allow users to plan their traffic shape scenario as they want using for example a go code that makes the playbook at the end.